lib/spack/spack/fetch_strategy.py

```diff
@@ -768,173 +896,216 @@ def __init__(self, **kwargs):
         self.submodules_delete = kwargs.get('submodules_delete', False)
         self.get_full_repo = kwargs.get('get_full_repo', False)
 
-    @property
-    def git_version(self):
-        vstring = self.git('--version', output=str).lstrip('git version ')
+        try:
+            self.ref = GitRef.from_directive(
+                commit=self.commit,
+                tag=self.tag,
+                branch=self.branch,
+                version_name=self.version_name)
+        except InvalidGitRef as e:
+            raise FetcherConflict(
```

Comment:
Is there a reason we change the type of the error here? Maybe I'll understand before I finish the review, but I'm not seeing it yet.

Reply:
`GitRef.from_directive()` does a few things:
1. It defines the precise inputs that make up the concept of a symbolic "git reference".
2. It applies a set of precedence rules which define the `version()` matching logic.
3. It provides a specific error describing how its inputs failed to create an unambiguous "git reference", if applicable.

None of those things correspond to a `FetcherConflict`. It is only when we attempt to *parse the `GitRef` from the `version()` directive kwargs* that the "ambiguous git ref" error becomes a "fetcher failure" error. This is also why I made `InvalidGitRef` subclass `ValueError` and put it at the top of the file, since it's *not* supposed to ever be raised outside of this module -- we should be raising a subclass of `SpackError` for that.

Reply:
Ok, after revising the thing a lot, the reason we now continue to change the type of the error is so that we can throw an `InvalidGitRef` exception without having to know the remote url. This means `GitRef.from_directive()` and other methods on `GitRef` can avoid having to pass in the url to construct their error strings.